noise/README.md

```diff
@@ -0,0 +1,400 @@
+# noise-libp2p - Channel Security Protocol
+
+> A libp2p transport security protocol built with the Noise Protocol Framework.
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+|-----------------|---------------|--------|-----------------|
+| 1A              | Working Draft | Active | r0, 2019-08-05  |
+
+Authors: [@yusefnapora]
+
+Interest Group: [@raulk], [@tomaka], [@romanb]
+
+[@yusefnapora]: https://github.com/yusefnapora
+[@raulk]: https://github.com/raulk
+[@tomaka]: https://github.com/tomaka
+[@romanb]: https://github.com/romanb
+
+See the [lifecycle document][lifecycle-spec] for context about maturity level
+and spec status.
+
+[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md
+
+## Overview
+
+The [Noise Protocol Framework][npf] is a framework for building security
+protocols by composing a small set of cryptographic primitives into patterns
+with verifiable security properties.
+
+This document specifies noise-libp2p, a libp2p transport security protocol built
+using the Noise Protocol Framework. As a framework for building protocols rather
+than a protocol itself, Noise presents a large decision space with many
+tradeoffs. The [Design Considerations
+section](#design-considerations-and-tradeoffs) covers the rationale behind the
+presented design.
+
+libp2p peers have an "identity keypair" which is used to derive their peer id,
+as described in the [peer id spec][peer-id-spec]. noise-libp2p uses the identity
+keypair to authenticate a separate static key which is used only for
+noise-libp2p. This is described in the [Static Key Authentication
+section](#static-key-authentication).
+
+## Message Framing
+
+noise-libp2p sends all data over the wire in the form of Noise messages, which
+are [defined][npf-message-format] with a maximum length of 65535 bytes.
+
+Noise messages are sent prefixed with their length in bytes, encoded as an
+unsigned 16-bit integer in network (big-endian) byte order.
+
+## Supported DH Functions, Ciphers, and Hash Functions
+
+The Noise framework allows protocol designers to choose from a small set of
+Diffie-Hellman functions, symmetric ciphers, and hash functions.
+
+For simplicity, noise-libp2p currently supports one function from each category:
+
+### Supported DH Functions
+
+noise-libp2p supports the 25519 DH functions [as defined in the Noise
+spec][npf-dh-25519].
+
+### Supported Cipher Functions
+
+noise-libp2p supports the ChaChaPoly cipher functions [as defined in the Noise
+spec][npf-cipher-chachapoly].
```

Comment:
AESGCM is a lot faster when running on hardware that has AES-support. For TLS, @kubuxu came up with a [nifty trick](https://github.com/libp2p/go-libp2p-tls/blob/46e9e9f9b00e3f896a17b8613bfb6efde5b177cc/crypto.go#L202-L222) to select AES if both peers have hardware support.

Is there any reason to exclude AESGCM here?

Reply:
@marten-seemann it was mostly about selecting ChaCha/Salsa on hardware that doesn't support it hardware AES. Hardware AES vs Salsa family are very similar in speed.

Reply:
I remember that the difference was quite large. I did a quick Google search and found this article (unfortunately from 2014): https://www.zeitgeist.se/2014/08/23/optimize-aes-and-chacha20-usage-with-boringssl/. In that benchmark, AES beats ChaCha by a factor of 2, if the CPU has hardware support.

What's the reason for making ChaCha required and AES-GCM optional? Just for reference, that's the opposite of what [TLS1.3 is requiring](https://tools.ietf.org/html/rfc8446#section-9.1).

Reply:
I just went with what's currently in rust-libp2p and figured we'd sort it out in discussion. I think we probably should add AESGCM since hardware support is pretty pervasive.

 Is there a reason not to line up with TLS 1.3 and make AES required, with a recommendation to also support ChaCha? Seems fairly sensible.

Reply:
From my runs on i5-4200M (only one I have available right now), I'mne  getting:
```
BenchmarkCacha20Poly1305-4   	    3000	    460213 ns/op	1139.23 MB/s
BenchmarkAES128GCM-4         	    5000	    301244 ns/op	1740.41 MB/s
BenchmarkAES256GCM-4         	    5000	    347137 ns/op	1510.32 MB/s
```

Here is Go's benchmark: https://github.com/Kubuxu/go-crypto-bench/
It would best to compare with more powerful Intel and AMD processor and also some ARMs.

Reply:
@Kubuxu Thanks for the benchmark! I ran it on my Macbook, and it looks like the difference is a lot bigger here:
```
goos: darwin
goarch: amd64
BenchmarkChacha20Poly1305-12    	   30000	    321145 ns/op	1632.56 MB/s
BenchmarkAES128GCM-12           	  100000	    145071 ns/op	3614.00 MB/s
BenchmarkAES256GCM-12           	   50000	    156157 ns/op	3357.44 MB/s
```

Reply:
I'm not a real cryptographer (and I hope @burges can correct me where I'm wrong), but I have a strong preference of ChaCha-Poly constructions over AES-GCM, for the following reasons:

1. I think we should also at least benchmark Chacha-Poly vs AES on some smartphone CPUs. Those usually don't have AES-NI or comparable hardware acceleration, and I expect the results to be much in favor of Chacha (since it was benchmarked "in software" in your tests as well, and performs at least comparable to AES "in hardware"). Chacha is based on add-rotate-xor (ARX) PRNG, which can be relatively efficiently implemented in most hardware infrastructures.
I would pretty much expect that mobile (or even embedded and IoT) devices should be in scope for libp2p design choices?

2. Continuation of the previous point, we would require AES implementations "in software" which don't suck, and, surprise-surprise, those are rare — there're a lot of potential footguns with timing attacks and AES in general not being very _copiler-_ or _developer-friendly_. It's much easier to achieve constant-time execution for Chacha than for AES, and the implementation will be much easier (taking in account you probably shouldn't trust your compiler for the critical parts anyway, and should roll out your own platform-specific assembly to substitute for that; I would expect ARX in assembly to be much much simpler).

3. AFAIU there're a couple of cryptography-related footguns in AES-GCM as well, so AFAIR NIST doesn't recommend using it for more than 2³² messages with the same key, even if you use cryptographically-strong nonces, and do everything else correctly.
Not sure how much of a practical limitation it is for libp2p, but at least it worth pointing out.